pkg/operator/controller/dns/controller.go

```diff
@@ -415,6 +476,73 @@ func conditionChanged(a, b iov1.DNSZoneCondition) bool {
 	return a.Status != b.Status || a.Reason != b.Reason || a.Message != b.Message
 }
 
+// migrateRecordStatusConditions purges deprecated fields from DNS Record
+// status. Removes "Failed" (DNSRecordFailedConditionType) and replaces with
+// "Published" (DNSRecordPublishedConditionType). If the status needs to be
+// updated, this function performs the update.  The function returns a Boolean
+// value indicating whether it has performed an update along with the error
+// value from any update it has performed.
+//
+// This function is at least required for 1 minor release and can be removed after
+// Openshift 4.13 is released.
+func (r *reconciler) migrateRecordStatusConditions(record *iov1.DNSRecord) (bool, error) {
+	var updateConditions bool
+	for i := range record.Status.Zones {
+		zone := record.Status.DeepCopy().Zones[i]
+		updateZone, cond := migrateRecordStatusCondition(zone.Conditions)
+		if updateZone {
+			updateConditions = true
+			record.Status.Zones[i].Conditions = cond
+		}
+	}
+	if updateConditions {
+		if err := r.client.Status().Update(context.TODO(), record); err != nil {
+			return false, fmt.Errorf("failed to update dnsrecord status: %w", err)
+		}
+		log.Info("dnsrecord status migrated", "dnsrecord", record)
+		return true, nil
+	}
+	return false, nil
+}
+
+func migrateRecordStatusCondition(conditions []iov1.DNSZoneCondition) (bool, []iov1.DNSZoneCondition) {
+	var result []iov1.DNSZoneCondition
+	var updatedCondition iov1.DNSZoneCondition
+	var updated bool
+	now := metav1.NewTime(clock.Now())
+	for _, cond := range conditions {
+		if cond.Type != iov1.DNSRecordFailedConditionType {
+			updatedCondition = cond
+		} else {
+			updated = true
+			switch cond.Status {
+			case string(operatorv1.ConditionTrue):
+				updatedCondition = iov1.DNSZoneCondition{
+					Type:               iov1.DNSRecordPublishedConditionType,
+					Status:             string(operatorv1.ConditionFalse),
+					LastTransitionTime: now,
+					Reason:             cond.Reason,
+					Message:            cond.Message,
+				}
+			case string(operatorv1.ConditionFalse):
+				updatedCondition = iov1.DNSZoneCondition{
+					Type:               iov1.DNSRecordPublishedConditionType,
+					Status:             string(operatorv1.ConditionTrue),
+					LastTransitionTime: now,
+					Reason:             cond.Reason,
+					Message:            cond.Message,
+				}
+				// We don't need to map Failed=Unknown condition
+				// since the operator never set this condition to Unknown.
+				// Since no version of the operator had this condition, we
+				// don't need to migrate it.
+			}
+		}
+		result = mergeConditions(result, []iov1.DNSZoneCondition{updatedCondition})
```

Comment:
Won't this add a zero status condition for `Published=Unknown`?  I know we don't expect that status condition ever to be set, but I would still prefer to have a default case:
```suggestion
			switch cond.Status {
			case string(operatorv1.ConditionTrue):
				updatedCondition = iov1.DNSZoneCondition{
					Type:               iov1.DNSRecordPublishedConditionType,
					Status:             string(operatorv1.ConditionFalse),
					LastTransitionTime: now,
					Reason:             cond.Reason,
					Message:            cond.Message,
				}
			case string(operatorv1.ConditionFalse):
				updatedCondition = iov1.DNSZoneCondition{
					Type:               iov1.DNSRecordPublishedConditionType,
					Status:             string(operatorv1.ConditionTrue),
					LastTransitionTime: now,
					Reason:             cond.Reason,
					Message:            cond.Message,
				}
			default:
				// We don't need to map Failed=Unknown condition
				// since the operator never set this condition to Unknown.
				// Since no version of the operator had this condition, we
				// don't need to migrate it.
				continue
			}
		}
		result = mergeConditions(result, []iov1.DNSZoneCondition{updatedCondition})
```